### ISSUE TITLE
Bug: Inconsistent TSMappedType AST shape for optional and readonly

### ISSUE THREAD

**Issue by JLHwung** [typescript-eslint/typescript-eslint#11055]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug) and found none that matched my issue.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Relevant Package

ast-spec

### Playground Link

https://typescript-eslint.io/play/#ts=5.8.3&showAST=es&fileType=.ts&code=G4QwTgBCBcEN4QNoGsIEsB2UME8C6sIuEAvgFBmiQBGsCK6WR%2BA-IceZeBAMZ0RgApiAAmAewwAbHElSZs%2BdjM5UII-gGohoidNmMFeNgtIVVg-gFpt4qTIbzmBEySA&eslintrc=N4KABGBEBOCuA2BTAzpAXGYBfEWg&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=true

### Repro Code

```TypeScript
var a: { [k in any]: any }

var b: { [k in any]?: any }

var c: { readonly [k in any]: any }

var d: { +readonly [k in any]?: any }

var e: { -readonly [k in any]: any }
```

### ESLint Config

```javascript

```

### tsconfig

```jsonc

```

### Expected Result

The 1st TSMappedType should have properties `{ optional: false, readonly: undefined }`

The 2nd TSMappedType should have property `{ readonly: undefined }`

The 3rd TSMappedType should have property `{ optional: false }`

The 5th TSMappedType should have property `{ optional: false }`

### Actual Result

The aforementioned TSMappedTypes are missing such properties, resulted to inconsistent AST shapes.

### Versions

| package                                | version |
| -------------------------------------- | ------- |
| `@typescript-eslint/typescript-estree` | `8.29.1` |

**Comment by JoshuaKGoldberg** [comment]
#5020 - is there a reason the AST should have the falsy properties defined?

**Comment by JLHwung** [comment]
I believe this issue is exactly part of #5020:

For example, the second one `var b: { [k in any]?: any }` is represented by a `TSMappedType` with an `optional: true` property.

while the first one `var a: { [k in any]: any }` is represented by a `TSMappedType` _without_ the `optional` property. So that means the `optional` property is an optional AST properties of `TSMappedType`. The conclusion of #5020 is that 
> Having a standard shape should help V8 optimise things at the (allbeit small) memory impact of additional "useless" values.

So we should always define the `optional` property, ditto for the `readonly` property.

**Comment by bradzacher** [comment]
This is a bug in the translation logic yeah -- it should always define the boolean flags as `optional: false` and `readonly: false` -- we should never omit them.

**Comment by JLHwung** [comment]
Note that `readonly` is an enum of `true`, `+` and `-`. So previously I proposed `undefined` as the default value. I guess `false` does no harm here.
